crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -576,23 +582,66 @@ impl CallChainLayout {
     }
 }
 
-fn has_parentheses(expr: &Expr, source: &str) -> bool {
-    has_own_parentheses(expr) || is_expression_parenthesized(AnyNodeRef::from(expr), source)
+#[derive(Debug, Copy, Clone, PartialEq, Eq, is_macro::Is)]
+pub(crate) enum OwnParentheses {
+    /// The node has parentheses, but they are empty (e.g., `[]` or `f()`).
+    Empty,
+    /// The node has parentheses, and they are non-empty (e.g., `[1]` or `f(1)`).
+    NonEmpty,
 }
 
-pub(crate) const fn has_own_parentheses(expr: &Expr) -> bool {
-    matches!(
-        expr,
-        Expr::Dict(_)
-            | Expr::List(_)
-            | Expr::Tuple(_)
-            | Expr::Set(_)
-            | Expr::ListComp(_)
-            | Expr::SetComp(_)
-            | Expr::DictComp(_)
-            | Expr::Call(_)
-            | Expr::Subscript(_)
-    )
+fn has_parentheses(expr: &Expr, source: &str) -> Option<OwnParentheses> {
+    if is_expression_parenthesized(AnyNodeRef::from(expr), source) {
+        Some(OwnParentheses::NonEmpty)
+    } else {
+        has_own_parentheses(expr, source)
+    }
+}
+
+/// Returns `true` if an [`Expr`] has its own parentheses.
+///
+/// A node is considered to have its own parentheses if it includes a `[]`, `()`, or `{}` pair
+/// that is inherent to the node.
+pub(crate) fn has_own_parentheses(expr: &Expr, source: &str) -> Option<OwnParentheses> {
+    fn has_comments_in(item: AnyNodeRef, source: &str) -> bool {
+        SimpleTokenizer::new(source, item.range())
+            .any(|token| matches!(token.kind, SimpleTokenKind::Comment))
+    }
```

Comment:
I think we should use `comments.has_dangling_comments(item)` because this is our source of truth when it comes to formatting comments (`placement.rs` could decide to move some comments out, this check would then return the wrong result)

Reply:
Ahh, smart